// @alcalzone/pak/build/cjs/lib/package-managers/npm/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var npm_exports = {};
__export(npm_exports, {
  Npm: () => Npm
});
module.exports = __toCommonJS(npm_exports);
var fs = __toESM(require("node:fs/promises"), 1);
var import_ky = __toESM(require("ky"), 1);
var path = __toESM(require("path"), 1);
var import_package_manager = require("../package-manager.js");
var import_semver = require("semver");
const exactVersionRegex = /.+@\d+/;
async function resolveDependency(dependency, version) {
  let reg;
  try {
    reg = await import_ky.default.get(`https://registry.npmjs.org/${dependency}`).json();
  } catch (e) {
    throw new Error(`Failed to download package info from npm registry: ${e.message}`, { cause: e });
  }
  if (version in reg.versions) {
    const versionInfo = reg.versions[version];
    return {
      version: versionInfo.version,
      dependencies: versionInfo.dependencies,
      integrity: versionInfo.dist.integrity,
      tarball: versionInfo.dist.tarball
    };
  } else {
    throw new Error(`${dependency}@${version} was not found in the npm registry!`);
  }
}
__name(resolveDependency, "resolveDependency");
function overrideV1(original, override) {
  const ret = {
    ...original
  };
  ret.version = override.version;
  if (ret.tarball) {
    ret.tarball = override.tarball;
  } else {
    ret.resolved = override.tarball;
    if (override.integrity)
      ret.integrity = override.integrity;
  }
  ret.requires = override.dependencies;
  return ret;
}
__name(overrideV1, "overrideV1");
function overrideV2(original, override) {
  const ret = {
    ...original
  };
  ret.version = override.version;
  if (ret.tarball) {
    ret.tarball = override.tarball;
  } else {
    ret.resolved = override.tarball;
    if (override.integrity)
      ret.integrity = override.integrity;
    else
      delete ret.integrity;
  }
  ret.dependencies = override.dependencies;
  return ret;
}
__name(overrideV2, "overrideV2");
function walkLockfileV1(root, dir, overrides, affectedPackageJsons) {
  if (!("dependencies" in root))
    return;
  for (const dep of Object.keys(root.dependencies)) {
    if (dep in overrides) {
      root.dependencies[dep] = overrideV1(root.dependencies[dep], overrides[dep]);
    } else {
      const depRoot = root.dependencies[dep];
      if ("requires" in depRoot) {
        let wasChanged = false;
        for (const [ovrr, { version }] of Object.entries(overrides)) {
          if (ovrr in depRoot.requires) {
            depRoot.requires[ovrr] = version;
            wasChanged = true;
          }
        }
        if (wasChanged) {
          affectedPackageJsons.add(path.join(dir, "node_modules", dep, "package.json"));
        }
      }
      walkLockfileV1(root.dependencies[dep], path.join(dir, "node_modules", dep), overrides, affectedPackageJsons);
    }
  }
}
__name(walkLockfileV1, "walkLockfileV1");
function walkLockfileV2V3(root, dir, overrides, affectedPackageJsons) {
  if (!("packages" in root))
    return;
  for (const pkg of Object.keys(root.packages)) {
    const pkgRoot = root.packages[pkg];
    const name = pkgRoot.name ?? pkg.substr(pkg.lastIndexOf("node_modules/") + 13);
    if (name in overrides) {
      root.packages[pkg] = overrideV2(root.packages[pkg], overrides[name]);
    } else if ("dependencies" in pkgRoot) {
      let wasChanged = false;
      for (const [ovrr, { version }] of Object.entries(overrides)) {
        if (ovrr in pkgRoot.dependencies) {
          pkgRoot.dependencies[ovrr] = version;
          wasChanged = true;
        }
      }
      if (wasChanged) {
        affectedPackageJsons.add(path.join(dir, pkg, "package.json"));
      }
    }
  }
}
__name(walkLockfileV2V3, "walkLockfileV2V3");
class Npm extends import_package_manager.PackageManager {
  static {
    __name(this, "Npm");
  }
  /** Executes a "raw" npm command */
  command(args) {
    const isWindows = process.platform === "win32";
    return this.exec("npm", args, {
      // Windows needs shell to be true to execute npm
      // https://github.com/nodejs/node/releases/tag/v18.20.2
      shell: isWindows
    });
  }
  /** Installs the given packages using npm */
  install(packages = [], options = {}) {
    const args = ["install"];
    if (options.dependencyType === "dev") {
      args.push("--save-dev");
    }
    if (options.exact || packages.some((p) => exactVersionRegex.test(p))) {
      args.push("--save-exact");
    }
    if (options.global)
      args.push("--global");
    if (options.force)
      args.push("--force");
    if (this.loglevel) {
      args.push("--loglevel", this.loglevel);
    }
    args.push(...packages);
    if (packages.length === 0) {
      if (this.environment === "production") {
        args.push("--production");
      }
      if (options.ignoreScripts) {
        args.push("--ignore-scripts");
      }
    }
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  uninstall(packages, options = {}) {
    const args = ["uninstall"];
    if (options.dependencyType === "dev") {
      args.push("--save-dev");
    }
    if (options.global)
      args.push("--global");
    if (this.loglevel) {
      args.push("--loglevel", this.loglevel);
    }
    args.push(...packages);
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  update(packages = [], options = {}) {
    const args = ["update"];
    if (options.dependencyType === "dev") {
      args.push("--save-dev");
    }
    if (options.global)
      args.push("-g");
    if (this.loglevel) {
      args.push("--loglevel", this.loglevel);
    }
    args.push(...packages);
    return this.command(args);
  }
  rebuild(packages = []) {
    const args = ["rebuild"];
    if (this.loglevel) {
      args.push("--loglevel", this.loglevel);
    }
    args.push(...packages);
    return this.command(args);
  }
  async detect(requireLockfile = true, setCwdToPackageRoot = false) {
    try {
      const root = await this.findRoot(requireLockfile ? "package-lock.json" : void 0);
      if (setCwdToPackageRoot)
        this.cwd = root;
      return true;
    } catch {
      return false;
    }
  }
  async version() {
    const result = await this.command(["-v"]);
    if (!result.success) {
      throw new Error(`Could not detect npm version: ${result.stderr}! Exit code: ${result.exitCode}.`);
    }
    return result.stdout;
  }
  async overrideDependencies(dependencies) {
    let root;
    let rootPackageJsonPath;
    let rootPackageLockPath;
    let rootPackageLock;
    try {
      root = await this.findRoot("package-lock.json");
      rootPackageJsonPath = path.join(root, "package.json");
      rootPackageLockPath = path.join(root, "package-lock.json");
      rootPackageLock = JSON.parse(await fs.readFile(rootPackageLockPath, "utf8"));
    } catch (e) {
      return this.fail(`Error loading root package.json and package-lock.json: ${e.message}`);
    }
    if (rootPackageLock.lockfileVersion !== 1 && rootPackageLock.lockfileVersion !== 2 && rootPackageLock.lockfileVersion !== 3) {
      return this.fail(`Lockfile version ${rootPackageLock.lockfileVersion} is not supported!`);
    }
    const overrides = {};
    for (const [dep, version] of Object.entries(dependencies)) {
      try {
        overrides[dep] = await resolveDependency(dep, version);
      } catch (e) {
        return this.fail(e.message);
      }
    }
    const affectedPackageJsons = /* @__PURE__ */ new Set();
    affectedPackageJsons.add(rootPackageJsonPath);
    if (rootPackageLock.lockfileVersion === 2 || rootPackageLock.lockfileVersion === 3) {
      walkLockfileV2V3(rootPackageLock, root, overrides, affectedPackageJsons);
    }
    if (rootPackageLock.lockfileVersion <= 2) {
      walkLockfileV1(rootPackageLock, root, overrides, affectedPackageJsons);
    }
    try {
      for (const packPath of affectedPackageJsons) {
        const pack = JSON.parse(await fs.readFile(packPath, "utf8"));
        if (!pack.dependencies)
          continue;
        let wasChanged = false;
        for (const [dep, { version }] of Object.entries(overrides)) {
          if (dep in pack.dependencies) {
            pack.dependencies[dep] = version;
            wasChanged = true;
          }
        }
        if (wasChanged) {
          await fs.writeFile(packPath, JSON.stringify(pack, null, 2) + "\n", "utf8");
        }
      }
      await fs.writeFile(rootPackageLockPath, JSON.stringify(rootPackageLock, null, 2) + "\n", "utf8");
    } catch (e) {
      return this.fail(`Error updating package files: ${e.message}`);
    }
    const prevCwd = this.cwd;
    this.cwd = root;
    try {
      const ret = await this.install();
      await this.command(["dedupe"]);
      return ret;
    } finally {
      this.cwd = prevCwd;
    }
  }
  async pack({ targetDir = this.cwd, workspace = "." } = {}) {
    const workspaceDir = path.join(this.cwd, workspace);
    const npmVersion = await this.version();
    if ((0, import_semver.gte)(npmVersion, "7.0.0")) {
      await fs.mkdir(targetDir, { recursive: true });
      const prevCwd = this.cwd;
      this.cwd = workspaceDir;
      try {
        const result = await this.command([
          "pack",
          "--pack-destination",
          targetDir
        ]);
        return {
          ...result,
          // npm outputs the filename of the tarball as stdout, we want the full path
          stdout: path.join(targetDir, result.stdout)
        };
      } finally {
        this.cwd = prevCwd;
      }
    } else {
      if (workspace !== ".") {
        return this.fail(`npm ${npmVersion} does not support monorepos`);
      }
      await fs.mkdir(targetDir, { recursive: true });
      const packageJsonPath = path.join(workspaceDir, "package.json");
      const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
      const packageName = packageJson.name.replace("/", "-").replace(/^@/, "");
      const version = packageJson.version;
      const targetPath = path.join(targetDir, `${packageName}-${version}.tgz`);
      const prevCwd = this.cwd;
      this.cwd = workspaceDir;
      try {
        const result = await this.command(["pack"]);
        if (!result.success)
          return result;
        const npmTarballPath = path.join(workspaceDir, result.stdout);
        if (npmTarballPath !== targetPath) {
          await fs.rename(npmTarballPath, targetPath);
        }
        return {
          ...result,
          stdout: targetPath
        };
      } finally {
        this.cwd = prevCwd;
      }
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Npm
});
