// @iobroker/adapter-react/Components/404.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _styles = require("@material-ui/core/styles");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _i18n = _interopRequireDefault(require("../i18n"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var styles = function styles(theme) {
  return {
    content: Object.assign({
      background: 'black',
      position: 'absolute',
      top: 0,
      bottom: 0,
      left: 0,
      right: 0
    }, theme.content),
    root: {
      boxSizing: 'border-box',
      height: '100%',
      width: '100%',
      backgroundColor: '#000000',
      backgroundImage: 'radial-gradient(#104254 , #05181c)',
      fontFamily: "'Inconsolata', Helvetica, sans-serif",
      fontSize: '1.5rem',
      color: 'rgba(128, 175, 255, 0.8)',
      textShadow: "0 0 1ex rgba(51, 70, 255, 1),\n        0 0 2px rgba(255, 255, 255, 0.8)"
    },
    overlay: {
      pointerEvents: 'none',
      width: '100%',
      height: '100%',
      background: "repeating-linear-gradient(\n                180deg,\n            rgba(0, 0, 0, 0) 0,\n            rgba(0, 0, 0, 0.3) 50%,\n            rgba(0, 0, 0, 0) 100%)",
      backgroundSize: 'auto 4px',
      zIndex: 99,
      '&::before': {
        content: '""',
        pointerEvents: 'none',
        position: 'absolute',
        display: 'block',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        width: '100%',
        height: '100%',
        backgroundImage: "linear-gradient(\n                0deg,\n                transparent 0%,\n                rgba(32, 50, 128, 0.2) 2%,\n                rgba(32, 50, 128, 0.8) 3%,\n                rgba(32, 50, 128, 0.2) 3%,\n                transparent 100%)",
        backgroundRepeat: 'no-repeat',
        animation: 'scan 7.5s linear 0s infinite'
      }
    },
    terminal: {
      boxSizing: 'inherit',
      position: 'absolute',
      height: '100%',
      width: 'calc(100% - 8rem)',
      maxWidth: '100%',
      padding: '4rem',
      textTransform: 'uppercase'
    },
    output: {
      color: 'rgba(128, 175, 255, 0.8)',
      textShadow: "\n        0 0 1px rgba(51, 70, 255, 0.4),\n            0 0 2px rgba(255, 255, 255, 0.8)",
      '&::before': {
        content: '"> "'
      }
    },
    errorCode: {
      color: 'white'
    }
  };
};

var Page404 = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(Page404, _Component);

  var _super = _createSuper(Page404);

  function Page404() {
    (0, _classCallCheck2["default"])(this, Page404);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(Page404, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: this.props.classes.content
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: this.props.classes.root
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: this.props.classes.overlay
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: this.props.classes.terminal
      }, /*#__PURE__*/_react["default"].createElement("h1", null, "Error", /*#__PURE__*/_react["default"].createElement("span", {
        className: this.props.classes.errorCode
      }, "404")), /*#__PURE__*/_react["default"].createElement("p", {
        className: this.props.classes.output
      }, _i18n["default"].t('The page you are looking for')), /*#__PURE__*/_react["default"].createElement("p", {
        className: this.props.classes.output
      }, _i18n["default"].t('Good luck'))))));
    }
  }]);
  return Page404;
}(_react.Component);

Page404.propTypes = {
  language: _propTypes["default"].string,
  onNavigate: _propTypes["default"].func,
  theme: _propTypes["default"].string,
  mobile: _propTypes["default"].bool
};

var _default = (0, _styles.withStyles)(styles)(Page404);

exports["default"] = _default;
