// @iobroker/adapter-react/Components/Loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

require("./loader.css");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var styles = function styles(theme) {
  return {};
};
/**
 * @typedef {object} LoaderProps
 * @property {string} [key] The key to identify this component.
 * @property {number} [size] The size in pixels of this loader.
 * @property {string} [themeType] The chosen theme type.
 * @property {string} [theme] The chosen theme.
 *
 * @extends {React.Component<LoaderProps>}
 */


var Loader = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(Loader, _React$Component);

  var _super = _createSuper(Loader);

  /**
   * @param {LoaderProps} props
   */
  function Loader(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Loader);
    _this = _super.call(this, props);
    _this.size = _this.props.size || 234;
    return _this;
  }

  (0, _createClass2["default"])(Loader, [{
    key: "render",
    value: function render() {
      var theme = this.props.themeType || this.props.theme || 'light';
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-back logo-background-' + theme
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: "logo-div",
        style: {
          width: this.size,
          height: this.size
        }
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-top logo-background-' + theme,
        style: {
          left: '37%'
        }
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-top logo-background-' + theme,
        style: {
          left: '57%'
        }
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-border logo-background-' + theme + ' logo-animate-wait',
        style: {
          borderWidth: this.size * 0.132
        }
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-i logo-animate-color-inside-' + theme
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-i-top logo-animate-color-inside-' + theme,
        style: {
          top: '18%'
        }
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-i-top logo-animate-color-inside-' + theme,
        style: {
          bottom: '18%'
        }
      })), /*#__PURE__*/_react["default"].createElement("div", {
        className: 'logo-animate-grow logo-animate-grow-' + theme,
        style: {
          width: this.size + 11,
          height: this.size + 11
        }
      }));
    }
  }]);
  return Loader;
}(_react["default"].Component);

Loader.propTypes = {
  size: _propTypes["default"].number,
  themeType: _propTypes["default"].string
};
/** @type {typeof Loader} */

var _export = (0, _withStyles["default"])(styles)(Loader);

var _default = _export;
exports["default"] = _default;
