// @iobroker/adapter-react/Components/TabContainer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Utils = _interopRequireDefault(require("./Utils"));

var _Grid = _interopRequireDefault(require("@material-ui/core/Grid"));

var _Paper = _interopRequireDefault(require("@material-ui/core/Paper"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var styles = {
  root: {
    width: '100%',
    height: '100%'
  },
  overflowHidden: {
    overflow: 'hidden'
  },
  container: {
    height: '100%'
  }
};
/**
 * @typedef {object} TabContainerProps
 * @property {string} [key] The key to identify this component.
 * @property {number} [elevation] The elevation of the tab container.
 * @property {string} [overflow] Set to 'visible' show the overflow.
 * @property {{ [key in keyof styles]: string}} classes The styling class names.
 *
 * @extends {React.Component<TabContainerProps>}
 */

var TabContainer = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(TabContainer, _React$Component);

  var _super = _createSuper(TabContainer);

  function TabContainer() {
    (0, _classCallCheck2["default"])(this, TabContainer);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(TabContainer, [{
    key: "render",
    value: function render() {
      var classes = this.props.classes;
      return /*#__PURE__*/_react["default"].createElement(_Paper["default"], {
        elevation: !isNaN(this.props.elevation) ? this.props.elevation : 1,
        className: _Utils["default"].clsx(classes.root, (0, _defineProperty2["default"])({}, classes.overflowHidden, this.props.overflow !== 'visible'))
      }, /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
        container: true,
        direction: "column",
        wrap: "nowrap",
        className: classes.container
      }, this.props.children));
    }
  }]);
  return TabContainer;
}(_react["default"].Component);

TabContainer.propTypes = {
  elevation: _propTypes["default"].number,
  overflow: _propTypes["default"].string
};
/** @type {typeof TabContainer} */

var _export = (0, _withStyles["default"])(styles)(TabContainer);

var _default = _export;
exports["default"] = _default;
