// @iobroker/adapter-react/Components/ToggleThemeMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ToggleThemeMenu;

var _react = _interopRequireDefault(require("react"));

var _core = require("@material-ui/core");

var _Brightness = _interopRequireDefault(require("@material-ui/icons/Brightness4"));

var _Brightness2 = _interopRequireDefault(require("@material-ui/icons/Brightness5"));

var _Brightness3 = _interopRequireDefault(require("@material-ui/icons/Brightness6"));

var _Brightness4 = _interopRequireDefault(require("@material-ui/icons/Brightness7"));

function ToggleThemeMenu(_ref) {
  var themeName = _ref.themeName,
      toggleTheme = _ref.toggleTheme,
      t = _ref.t,
      className = _ref.className,
      style = _ref.style,
      size = _ref.size;
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className || undefined,
    style: style || undefined
  }, /*#__PURE__*/_react["default"].createElement(_core.Tooltip, {
    title: t('Change color theme')
  }, /*#__PURE__*/_react["default"].createElement(_core.IconButton, {
    onClick: function onClick() {
      return toggleTheme();
    },
    size: size || 'medium'
  }, themeName === 'dark' && /*#__PURE__*/_react["default"].createElement(_Brightness["default"], {
    className: className
  }), themeName === 'blue' && /*#__PURE__*/_react["default"].createElement(_Brightness2["default"], {
    className: className
  }), themeName === 'colored' && /*#__PURE__*/_react["default"].createElement(_Brightness3["default"], {
    className: className
  }), themeName === 'light' && /*#__PURE__*/_react["default"].createElement(_Brightness4["default"], {
    className: className
  }))));
}
