// @alcalzone/pak/build/cjs/lib/package-managers/yarn-berry/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var yarn_berry_exports = {};
__export(yarn_berry_exports, {
  YarnBerry: () => YarnBerry
});
module.exports = __toCommonJS(yarn_berry_exports);
var fs = __toESM(require("node:fs/promises"), 1);
var path = __toESM(require("path"), 1);
var import_package_manager = require("../package-manager.js");
function setLoglevel(_args, _loglevel) {
}
__name(setLoglevel, "setLoglevel");
class YarnBerry extends import_package_manager.PackageManager {
  static {
    __name(this, "YarnBerry");
  }
  /** Executes a "raw" yarn command */
  command(args) {
    return this.exec("yarn", args);
  }
  /**
   * Installs the given packages using `yarn add`. If no packages are given, `yarn install` is executed in the cwd.
   */
  install(packages = [], options = {}) {
    const args = [];
    if (options.global) {
      return this.fail(`yarn berry does not support global installs!`);
    }
    if (packages.length > 0) {
      args.push("add");
      args.push(...packages);
      if (options.dependencyType === "dev") {
        args.push("--dev");
      }
      if (options.exact) {
        args.push("--exact");
      }
    } else {
      args.push("install");
    }
    setLoglevel(args, this.loglevel);
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  uninstall(packages, options = {}) {
    const args = [];
    if (options.global) {
      return this.fail(`yarn berry does not support global uninstalls!`);
    }
    args.push("remove");
    args.push(...packages);
    setLoglevel(args, this.loglevel);
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  update(packages = [], options = {}) {
    return this.install(packages, options);
  }
  rebuild(packages = []) {
    const args = ["rebuild"];
    args.push(...packages);
    setLoglevel(args, this.loglevel);
    return this.command(args);
  }
  async detect(requireLockfile = true, setCwdToPackageRoot = false) {
    try {
      const root = await this.findRoot(requireLockfile ? "yarn.lock" : void 0);
      const version = await this.version();
      if (!version.startsWith("2.") && !version.startsWith("3.") && !version.startsWith("4.")) {
        return false;
      }
      if (setCwdToPackageRoot)
        this.cwd = root;
      return true;
    } catch {
      return false;
    }
  }
  async version() {
    const result = await this.command(["-v"]);
    if (!result.success) {
      throw new Error(`Could not detect yarn version: ${result.stderr}! Exit code: ${result.exitCode}.`);
    }
    return result.stdout;
  }
  async overrideDependencies(dependencies) {
    let root;
    try {
      root = await this.findRoot("yarn.lock");
      const packageJsonPath = path.join(root, "package.json");
      const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
      let resolutions = packageJson.resolutions ?? {};
      resolutions = {
        ...resolutions,
        ...dependencies
      };
      packageJson.resolutions = resolutions;
      await fs.writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2) + "\n", "utf8");
    } catch (e) {
      const stderr = "Error updating root package.json: " + e.message;
      return {
        success: false,
        exitCode: 1,
        stdout: "",
        stderr,
        stdall: stderr
      };
    }
    const prevCwd = this.cwd;
    this.cwd = root;
    try {
      return await this.install();
    } finally {
      this.cwd = prevCwd;
    }
  }
  async pack({ targetDir = this.cwd, workspace = "." } = {}) {
    const workspaceDir = path.join(this.cwd, workspace);
    await fs.mkdir(targetDir, { recursive: true });
    const packageJsonPath = path.join(workspaceDir, "package.json");
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
    const packageName = packageJson.name.replace("/", "-").replace(/^@/, "");
    const version = packageJson.version;
    const targetPath = path.join(targetDir, `${packageName}-${version}.tgz`);
    const prevCwd = this.cwd;
    this.cwd = workspaceDir;
    try {
      const result = await this.command(["pack", "--out", targetPath]);
      return {
        ...result,
        // Yarn outputs a lot, not only the target path of the file. Manually return the absolute path.
        stdout: targetPath
      };
    } finally {
      this.cwd = prevCwd;
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  YarnBerry
});
