// @iobroker/adapter-react/Dialogs/Error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _Dialog = _interopRequireDefault(require("@material-ui/core/Dialog"));

var _DialogActions = _interopRequireDefault(require("@material-ui/core/DialogActions"));

var _DialogContent = _interopRequireDefault(require("@material-ui/core/DialogContent"));

var _DialogContentText = _interopRequireDefault(require("@material-ui/core/DialogContentText"));

var _DialogTitle = _interopRequireDefault(require("@material-ui/core/DialogTitle"));

var _Check = _interopRequireDefault(require("@material-ui/icons/Check"));

var _i18n = _interopRequireDefault(require("../i18n"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var styles = function styles(theme) {
  return {
    titleBackground: {},
    titleColor: {}
  };
};
/**
 * @typedef {object} DialogErrorProps
 * @property {string} [key] The key to identify this component.
 * @property {string} [title] The dialog title; default: Error (translated)
 * @property {string | JSX.Element} text The dialog text.
 * @property {() => void} [onClose] Close handler.
 * @property {{titleBackground: string; titleColor: string}} classes The styling class names.
 *
 * @extends {React.Component<DialogErrorProps>}
 */


var DialogError = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(DialogError, _React$Component);

  var _super = _createSuper(DialogError);

  function DialogError() {
    (0, _classCallCheck2["default"])(this, DialogError);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(DialogError, [{
    key: "handleOk",
    value: function handleOk() {
      this.props.onClose && this.props.onClose();
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      return /*#__PURE__*/_react["default"].createElement(_Dialog["default"], {
        open: true,
        maxWidth: "sm",
        fullWidth: true,
        onClose: function onClose() {
          return _this.handleOk();
        },
        "aria-labelledby": "alert-dialog-title",
        "aria-describedby": "alert-dialog-description"
      }, /*#__PURE__*/_react["default"].createElement(_DialogTitle["default"], {
        className: this.props.classes.titleBackground,
        classes: {
          root: this.props.classes.titleColor
        },
        id: "alert-dialog-title"
      }, this.props.title || _i18n["default"].t('ra_Error')), /*#__PURE__*/_react["default"].createElement(_DialogContent["default"], null, /*#__PURE__*/_react["default"].createElement(_DialogContentText["default"], {
        id: "alert-dialog-description"
      }, this.props.text || _i18n["default"].t('ra_Unknown error!'))), /*#__PURE__*/_react["default"].createElement(_DialogActions["default"], null, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
        variant: "contained",
        onClick: function onClick() {
          return _this.handleOk();
        },
        color: "primary",
        autoFocus: true,
        startIcon: /*#__PURE__*/_react["default"].createElement(_Check["default"], null)
      }, _i18n["default"].t('ra_Ok'))));
    }
  }]);
  return DialogError;
}(_react["default"].Component);

DialogError.propTypes = {
  onClose: _propTypes["default"].func,
  title: _propTypes["default"].string,
  text: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].element]),
  icon: _propTypes["default"].object
};
/** @type {typeof DialogError} */

var _export = (0, _withStyles["default"])(styles)(DialogError);

var _default = _export;
exports["default"] = _default;
