// @alcalzone/pak/build/cjs/lib/package-managers/yarn-classic/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var yarn_classic_exports = {};
__export(yarn_classic_exports, {
  YarnClassic: () => YarnClassic
});
module.exports = __toCommonJS(yarn_classic_exports);
var fs = __toESM(require("node:fs/promises"), 1);
var path = __toESM(require("path"), 1);
var import_package_manager = require("../package-manager.js");
function setLoglevel(args, loglevel) {
  if (loglevel === "silent" || loglevel === "verbose") {
    args.push(`--${loglevel}`);
  }
}
__name(setLoglevel, "setLoglevel");
class YarnClassic extends import_package_manager.PackageManager {
  static {
    __name(this, "YarnClassic");
  }
  /** Executes a "raw" yarn command */
  command(args) {
    return this.exec("yarn", args);
  }
  /**
   * Installs the given packages using `yarn add`. If no packages are given, `yarn install` is executed in the cwd.
   */
  install(packages = [], options = {}) {
    const args = [];
    if (packages.length > 0) {
      if (options.global)
        args.push("global");
      args.push("add");
      args.push(...packages);
      if (options.dependencyType === "dev") {
        args.push("--dev");
      }
      if (options.exact) {
        args.push("--exact");
      }
    } else {
      args.push("install");
      if (this.environment === "production") {
        args.push("--production");
      }
      if (options.ignoreScripts) {
        args.push("--ignore-scripts");
      }
      if (options.force) {
        args.push("--force");
      }
    }
    setLoglevel(args, this.loglevel);
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  uninstall(packages, options = {}) {
    const args = [];
    if (options.global)
      args.push("global");
    args.push("remove");
    args.push(...packages);
    if (options.dependencyType === "dev") {
      args.push("--dev");
    }
    setLoglevel(args, this.loglevel);
    if (options.additionalArgs) {
      args.push(...options.additionalArgs);
    }
    return this.command(args);
  }
  update(packages = [], options = {}) {
    return this.install(packages, options);
  }
  async rebuild() {
    const stderr = `yarn classic does not support the "rebuild" command!`;
    return {
      success: false,
      exitCode: 1,
      stdout: "",
      stderr,
      stdall: stderr
    };
  }
  async detect(requireLockfile = true, setCwdToPackageRoot = false) {
    try {
      const root = await this.findRoot(requireLockfile ? "yarn.lock" : void 0);
      if (!(await this.version()).startsWith("1."))
        return false;
      if (setCwdToPackageRoot)
        this.cwd = root;
      return true;
    } catch {
      return false;
    }
  }
  async version() {
    const result = await this.command(["-v"]);
    if (!result.success) {
      throw new Error(`Could not detect yarn version: ${result.stderr}! Exit code: ${result.exitCode}.`);
    }
    return result.stdout;
  }
  async overrideDependencies(dependencies) {
    let root;
    try {
      root = await this.findRoot("yarn.lock");
      const packageJsonPath = path.join(root, "package.json");
      const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
      let resolutions = packageJson.resolutions ?? {};
      resolutions = {
        ...resolutions,
        ...dependencies
      };
      packageJson.resolutions = resolutions;
      await fs.writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2) + "\n", "utf8");
    } catch (e) {
      const stderr = "Error updating root package.json: " + e.message;
      return {
        success: false,
        exitCode: 1,
        stdout: "",
        stderr,
        stdall: stderr
      };
    }
    const prevCwd = this.cwd;
    this.cwd = root;
    try {
      return await this.install();
    } finally {
      this.cwd = prevCwd;
    }
  }
  pack(_options) {
    return this.fail(`yarn classic does not support packing tarballs!`);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  YarnClassic
});
