// @alcalzone/pak/build/esm/lib/package-managers/package-manager.js
import glob from "tiny-glob";
import spawn from "nano-spawn";
import * as fs from "node:fs/promises";
import path from "path";
async function fileExists(filePath) {
    try {
        await fs.access(filePath);
        return true;
    }
    catch {
        return false;
    }
}
export class PackageManager {
    /** Executes a command and translates the result into a CommandResult */
    async exec(file, args, options = {}) {
        const subprocess = spawn(file, args, {
            ...options,
            cwd: this.cwd,
        });
        const childProcess = await subprocess.nodeChildProcess;
        // Pipe command outputs if desired
        if (this.stdout)
            childProcess.stdout?.pipe(this.stdout, { end: false });
        if (this.stderr)
            childProcess.stderr?.pipe(this.stderr, { end: false });
        if (this.stdall) {
            childProcess.stdout?.pipe(this.stdall, { end: false });
            childProcess.stderr?.pipe(this.stdall, { end: false });
        }
        try {
            const result = await subprocess;
            return {
                success: true,
                exitCode: 0,
                stdout: result.stdout,
                stderr: result.stderr,
                stdall: result.output,
            };
        }
        catch (error) {
            return {
                success: false,
                exitCode: error.exitCode ?? 1,
                stdout: error.stdout ?? "",
                stderr: error.stderr ?? "",
                stdall: error.output ?? "",
            };
        }
        finally {
            childProcess.stdout?.unpipe();
            childProcess.stderr?.unpipe();
        }
    }
    fail(message) {
        const stderr = message;
        return Promise.resolve({
            success: false,
            exitCode: 1,
            stdout: "",
            stderr,
            stdall: stderr,
        });
    }
    /** Finds the closest parent directory that contains a package.json and the corresponding lockfile (if one was specified) */
    async findRoot(lockfile) {
        let curDir = this.cwd;
        let parentDir;
        while (true) {
            const packageJsonPath = path.join(curDir, "package.json");
            if (await fileExists(packageJsonPath)) {
                if (!lockfile)
                    return curDir;
                const lockfilePath = path.join(curDir, lockfile);
                if (await fileExists(lockfilePath))
                    return curDir;
            }
            parentDir = path.dirname(curDir);
            if (parentDir === curDir) {
                // we've reached the root without finiding a package.json
                throw new Error(`This directory tree does not contain a directory with package.json${lockfile ? " and a lockfile" : ""}!`);
            }
            curDir = parentDir;
        }
    }
    /** Resolves absolute paths of all workspaces defined in the current monorepo */
    async workspaces() {
        const packageJsonPath = path.join(this.cwd, "package.json");
        const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
        const workspaces = packageJson.workspaces;
        if (!workspaces?.length)
            return [];
        // Workspaces are globs we need to resolve
        const ret = new Set();
        for (const ws of workspaces) {
            const resolved = await glob(ws, {
                cwd: this.cwd,
                absolute: true,
            });
            // Check every potential directory if it contains a package.json
            for (const r of resolved) {
                const workspacePackageJsonPath = path.join(r, "package.json");
                try {
                    if ((await fs.stat(workspacePackageJsonPath)).isFile()) {
                        ret.add(r);
                    }
                }
                catch {
                    // Does not exist, ignore
                }
            }
        }
        return [...ret].sort();
    }
    /** The directory to run the package manager commands in */
    cwd = process.cwd();
    /** Which loglevel to pass to the package manager */
    loglevel;
    /** The (optional) stream to pipe the command's stdout into */
    stdout;
    /** The (optional) stream to pipe the command's stderr into */
    stderr;
    /**
     * The (optional) stream to pipe the command's entire output (stdout + stderr) into.
     * If this is set, stdout and stderr will be ignored
     */
    stdall;
    /**
     * The environment the package manager is executed in (default: "production").
     * In an production environment, `pak` avoids accidentally pulling in `devDependencies`.
     */
    environment = "production";
}
