// @alcalzone/pak/build/cjs/lib/package-managers/package-manager.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var package_manager_exports = {};
__export(package_manager_exports, {
  PackageManager: () => PackageManager
});
module.exports = __toCommonJS(package_manager_exports);
var import_tiny_glob = __toESM(require("tiny-glob"), 1);
var import_nano_spawn = __toESM(require("nano-spawn"), 1);
var fs = __toESM(require("node:fs/promises"), 1);
var import_path = __toESM(require("path"), 1);
async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}
__name(fileExists, "fileExists");
class PackageManager {
  static {
    __name(this, "PackageManager");
  }
  /** Executes a command and translates the result into a CommandResult */
  async exec(file, args, options = {}) {
    const subprocess = (0, import_nano_spawn.default)(file, args, {
      ...options,
      cwd: this.cwd
    });
    const childProcess = await subprocess.nodeChildProcess;
    if (this.stdout)
      childProcess.stdout?.pipe(this.stdout, { end: false });
    if (this.stderr)
      childProcess.stderr?.pipe(this.stderr, { end: false });
    if (this.stdall) {
      childProcess.stdout?.pipe(this.stdall, { end: false });
      childProcess.stderr?.pipe(this.stdall, { end: false });
    }
    try {
      const result = await subprocess;
      return {
        success: true,
        exitCode: 0,
        stdout: result.stdout,
        stderr: result.stderr,
        stdall: result.output
      };
    } catch (error) {
      return {
        success: false,
        exitCode: error.exitCode ?? 1,
        stdout: error.stdout ?? "",
        stderr: error.stderr ?? "",
        stdall: error.output ?? ""
      };
    } finally {
      childProcess.stdout?.unpipe();
      childProcess.stderr?.unpipe();
    }
  }
  fail(message) {
    const stderr = message;
    return Promise.resolve({
      success: false,
      exitCode: 1,
      stdout: "",
      stderr,
      stdall: stderr
    });
  }
  /** Finds the closest parent directory that contains a package.json and the corresponding lockfile (if one was specified) */
  async findRoot(lockfile) {
    let curDir = this.cwd;
    let parentDir;
    while (true) {
      const packageJsonPath = import_path.default.join(curDir, "package.json");
      if (await fileExists(packageJsonPath)) {
        if (!lockfile)
          return curDir;
        const lockfilePath = import_path.default.join(curDir, lockfile);
        if (await fileExists(lockfilePath))
          return curDir;
      }
      parentDir = import_path.default.dirname(curDir);
      if (parentDir === curDir) {
        throw new Error(`This directory tree does not contain a directory with package.json${lockfile ? " and a lockfile" : ""}!`);
      }
      curDir = parentDir;
    }
  }
  /** Resolves absolute paths of all workspaces defined in the current monorepo */
  async workspaces() {
    const packageJsonPath = import_path.default.join(this.cwd, "package.json");
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, "utf8"));
    const workspaces = packageJson.workspaces;
    if (!workspaces?.length)
      return [];
    const ret = /* @__PURE__ */ new Set();
    for (const ws of workspaces) {
      const resolved = await (0, import_tiny_glob.default)(ws, {
        cwd: this.cwd,
        absolute: true
      });
      for (const r of resolved) {
        const workspacePackageJsonPath = import_path.default.join(r, "package.json");
        try {
          if ((await fs.stat(workspacePackageJsonPath)).isFile()) {
            ret.add(r);
          }
        } catch {
        }
      }
    }
    return [...ret].sort();
  }
  /** The directory to run the package manager commands in */
  cwd = process.cwd();
  /** Which loglevel to pass to the package manager */
  loglevel;
  /** The (optional) stream to pipe the command's stdout into */
  stdout;
  /** The (optional) stream to pipe the command's stderr into */
  stderr;
  /**
   * The (optional) stream to pipe the command's entire output (stdout + stderr) into.
   * If this is set, stdout and stderr will be ignored
   */
  stdall;
  /**
   * The environment the package manager is executed in (default: "production").
   * In an production environment, `pak` avoids accidentally pulling in `devDependencies`.
   */
  environment = "production";
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PackageManager
});
