// @iobroker/adapter-react/Components/Loaders/Vendor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _CircularProgress = _interopRequireDefault(require("@material-ui/core/CircularProgress"));

require("./Vendor.css");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

/**
 * @typedef {object} LoaderProps
 * @property {string} [key] The key to identify this component.
 * @property {number} [size] The size in pixels of this loader.
 * @property {string} [themeType] The chosen theme type.
 * @property {string} [theme] The chosen theme.
 *
 * @extends {React.Component<LoaderProps>}
 */
var LoaderVendor = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(LoaderVendor, _React$Component);

  var _super = _createSuper(LoaderVendor);

  /**
   * @param {LoaderProps} props
   */
  function LoaderVendor(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, LoaderVendor);
    _this = _super.call(this, props);
    _this.size = _this.props.size || 200;
    return _this;
  }

  (0, _createClass2["default"])(LoaderVendor, [{
    key: "render",
    value: function render() {
      var theme = this.props.themeType || this.props.theme || 'light';
      return /*#__PURE__*/_react["default"].createElement("div", {
        className: 'vendor-logo-back logo-background-' + theme,
        style: {
          display: 'flex',
          flexDirection: 'column',
          height: '100%',
          width: '10%',
          margin: 'auto'
        }
      }, /*#__PURE__*/_react["default"].createElement("div", {
        style: {
          flexGrow: 1
        }
      }), /*#__PURE__*/_react["default"].createElement(_CircularProgress["default"], {
        color: "secondary",
        size: 200,
        thickness: 5
      }), /*#__PURE__*/_react["default"].createElement("div", {
        style: {
          flexGrow: 1
        }
      }));
    }
  }]);
  return LoaderVendor;
}(_react["default"].Component);

LoaderVendor.propTypes = {
  size: _propTypes["default"].number,
  themeType: _propTypes["default"].string
};
/** @type {typeof LoaderVendor} */

var _export = LoaderVendor;
var _default = _export;
exports["default"] = _default;
